{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# To format dataframe for input of Neural Network \n",
    "\n",
    "* Initial testing to establish approach to apply to end datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import ast\n",
    "import random\n",
    "import string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"nnPrepFirstData.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homo sapiens (Human)                                                                                     2711\n",
       "Mus musculus (Mouse)                                                                                      767\n",
       "Escherichia coli (strain K12)                                                                             635\n",
       "Saccharomyces cerevisiae (strain ATCC 204508 / S288c) (Baker's yeast)                                     613\n",
       "Bacillus subtilis (strain 168)                                                                            283\n",
       "                                                                                                         ... \n",
       "Hepatitis C virus genotype 2b (isolate HC-J8) (HCV)                                                         1\n",
       "Pseudothermotoga lettingae (strain ATCC BAA-301 / DSM 14385 / NBR107922 / TMO) (Thermotoga lettingae)       1\n",
       "Thermus thermophilus (strain SG0.5JP17-16)                                                                  1\n",
       "Robinia pseudoacacia (Black locust)                                                                         1\n",
       "Magnaporthe grisea (Crabgrass-specific blast fungus) (Pyricularigrisea)                                     1\n",
       "Name: organism, Length: 3121, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['organism'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homo sapiens (Human)                                                                                     2711\n",
       "Mus musculus (Mouse)                                                                                      767\n",
       "Escherichia coli (strain K12)                                                                             635\n",
       "Saccharomyces cerevisiae (strain ATCC 204508 / S288c) (Baker's yeast)                                     613\n",
       "Bacillus subtilis (strain 168)                                                                            283\n",
       "                                                                                                         ... \n",
       "Hepatitis C virus genotype 2b (isolate HC-J8) (HCV)                                                         1\n",
       "Pseudothermotoga lettingae (strain ATCC BAA-301 / DSM 14385 / NBR107922 / TMO) (Thermotoga lettingae)       1\n",
       "Thermus thermophilus (strain SG0.5JP17-16)                                                                  1\n",
       "Robinia pseudoacacia (Black locust)                                                                         1\n",
       "Magnaporthe grisea (Crabgrass-specific blast fungus) (Pyricularigrisea)                                     1\n",
       "Name: organism, Length: 3121, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['organism'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assigning a unique identifier to each protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Correct Index</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncbi_taxon_id</th>\n",
       "      <th>shortened_sequence</th>\n",
       "      <th>Sequence_Length</th>\n",
       "      <th>acc</th>\n",
       "      <th>mobidb_consensus</th>\n",
       "      <th>shortened_mobidb_consensus</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>#_AA_referenced</th>\n",
       "      <th>Updated Indexes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>9606</td>\n",
       "      <td>AAPPKAVLKLEPQWINVLQEDSVTLTCRGTHSPESDSIQWFHNGNL...</td>\n",
       "      <td>173</td>\n",
       "      <td>P31994</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[46, 46...</td>\n",
       "      <td>[[45, 45, 'D'], [46, 49, 'C'], [50, 66, 'S'], ...</td>\n",
       "      <td>310</td>\n",
       "      <td>173</td>\n",
       "      <td>[[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Kluyveromyces lactis (strain ATCC 8585 / CBS 2...</td>\n",
       "      <td>284590</td>\n",
       "      <td>MGKGKPRGLNSARKLRVHRRNNRWAETTYKKRLLGTAFKSSPFGGS...</td>\n",
       "      <td>145</td>\n",
       "      <td>F2Z602</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[2, 143...</td>\n",
       "      <td>[[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>[[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Kluyveromyces lactis (Yeast) (Candida sphaerica)</td>\n",
       "      <td>28985</td>\n",
       "      <td>MGKGKPRGLNSARKLRVHRRNNRWAETTYKKRLLGTAFKSSPFGGS...</td>\n",
       "      <td>145</td>\n",
       "      <td>Q875M3</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 145...</td>\n",
       "      <td>[[1, 145, 'S']]</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>[[1, 145, 'S']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>10090</td>\n",
       "      <td>APYGADTPTACCFSYSRKIPRQFIVDYFETSSLCSQPGVIFLTKRN...</td>\n",
       "      <td>69</td>\n",
       "      <td>P10855</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[30, 91...</td>\n",
       "      <td>[[24, 29, 'D'], [30, 91, 'S'], [92, 92, 'D']]</td>\n",
       "      <td>92</td>\n",
       "      <td>69</td>\n",
       "      <td>[[1, 6, 'D'], [7, 68, 'S'], [69, 69, 'D']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>Dickeya chrysanthemi (Pectobacterium chrysanth...</td>\n",
       "      <td>556</td>\n",
       "      <td>ADKLPNIVILATGGTIAGSAATGTQTTGYKAGALGVDTLINAVPEV...</td>\n",
       "      <td>327</td>\n",
       "      <td>P06608</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[22, 22...</td>\n",
       "      <td>[[22, 24, 'C'], [25, 38, 'S'], [39, 55, 'C'], ...</td>\n",
       "      <td>348</td>\n",
       "      <td>327</td>\n",
       "      <td>[[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  Correct Index  \\\n",
       "0           0              0   \n",
       "1           1              1   \n",
       "2           2              2   \n",
       "3           3              4   \n",
       "4           4              5   \n",
       "\n",
       "                                            organism  ncbi_taxon_id  \\\n",
       "0                               Homo sapiens (Human)           9606   \n",
       "1  Kluyveromyces lactis (strain ATCC 8585 / CBS 2...         284590   \n",
       "2   Kluyveromyces lactis (Yeast) (Candida sphaerica)          28985   \n",
       "3                               Mus musculus (Mouse)          10090   \n",
       "4  Dickeya chrysanthemi (Pectobacterium chrysanth...            556   \n",
       "\n",
       "                                  shortened_sequence  Sequence_Length     acc  \\\n",
       "0  AAPPKAVLKLEPQWINVLQEDSVTLTCRGTHSPESDSIQWFHNGNL...              173  P31994   \n",
       "1  MGKGKPRGLNSARKLRVHRRNNRWAETTYKKRLLGTAFKSSPFGGS...              145  F2Z602   \n",
       "2  MGKGKPRGLNSARKLRVHRRNNRWAETTYKKRLLGTAFKSSPFGGS...              145  Q875M3   \n",
       "3  APYGADTPTACCFSYSRKIPRQFIVDYFETSSLCSQPGVIFLTKRN...               69  P10855   \n",
       "4  ADKLPNIVILATGGTIAGSAATGTQTTGYKAGALGVDTLINAVPEV...              327  P06608   \n",
       "\n",
       "                                    mobidb_consensus  \\\n",
       "0  {'disorder': {'derived': [{'regions': [[46, 46...   \n",
       "1  {'disorder': {'derived': [{'regions': [[2, 143...   \n",
       "2  {'disorder': {'derived': [{'regions': [[1, 145...   \n",
       "3  {'disorder': {'derived': [{'regions': [[30, 91...   \n",
       "4  {'disorder': {'derived': [{'regions': [[22, 22...   \n",
       "\n",
       "                          shortened_mobidb_consensus  Sequence Length  \\\n",
       "0  [[45, 45, 'D'], [46, 49, 'C'], [50, 66, 'S'], ...              310   \n",
       "1      [[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]              145   \n",
       "2                                    [[1, 145, 'S']]              145   \n",
       "3      [[24, 29, 'D'], [30, 91, 'S'], [92, 92, 'D']]               92   \n",
       "4  [[22, 24, 'C'], [25, 38, 'S'], [39, 55, 'C'], ...              348   \n",
       "\n",
       "   #_AA_referenced                                    Updated Indexes  \n",
       "0              173  [[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...  \n",
       "1              145      [[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]  \n",
       "2              145                                    [[1, 145, 'S']]  \n",
       "3               69         [[1, 6, 'D'], [7, 68, 'S'], [69, 69, 'D']]  \n",
       "4              327  [[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_list = []\n",
    "\n",
    "random.seed(18091)\n",
    "\n",
    "for rows in df.itertuples(): # iterating through each row in datafrae\n",
    "    x = ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(6)) # Creating 6 digit identifier containing letters and numbers\n",
    "    id_list.append(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Identifier'] = id_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Replacing each letter with a one hot encoding representation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('A','1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('B','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('C','0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('D','0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('E','0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('F','0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('G','0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('H','0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('I','0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('J','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('K','0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('L','0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('M','0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('N','0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('O','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('P','0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('Q','0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('R','0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('S','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('T','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('U','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('V','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('W','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('X','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('Y','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 ')\n",
    "df['shortened_sequence'] = df['shortened_sequence'].str.replace('Z','0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Correct Index</th>\n",
       "      <th>organism</th>\n",
       "      <th>ncbi_taxon_id</th>\n",
       "      <th>shortened_sequence</th>\n",
       "      <th>Sequence_Length</th>\n",
       "      <th>acc</th>\n",
       "      <th>mobidb_consensus</th>\n",
       "      <th>shortened_mobidb_consensus</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>#_AA_referenced</th>\n",
       "      <th>Updated Indexes</th>\n",
       "      <th>Identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>9606</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 ...</td>\n",
       "      <td>173</td>\n",
       "      <td>P31994</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[46, 46...</td>\n",
       "      <td>[[45, 45, 'D'], [46, 49, 'C'], [50, 66, 'S'], ...</td>\n",
       "      <td>310</td>\n",
       "      <td>173</td>\n",
       "      <td>[[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...</td>\n",
       "      <td>JV1H0X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Kluyveromyces lactis (strain ATCC 8585 / CBS 2...</td>\n",
       "      <td>284590</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>145</td>\n",
       "      <td>F2Z602</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[2, 143...</td>\n",
       "      <td>[[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>[[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]</td>\n",
       "      <td>12KUMW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>Kluyveromyces lactis (Yeast) (Candida sphaerica)</td>\n",
       "      <td>28985</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>145</td>\n",
       "      <td>Q875M3</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 145...</td>\n",
       "      <td>[[1, 145, 'S']]</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>[[1, 145, 'S']]</td>\n",
       "      <td>IXJPFB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>10090</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>69</td>\n",
       "      <td>P10855</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[30, 91...</td>\n",
       "      <td>[[24, 29, 'D'], [30, 91, 'S'], [92, 92, 'D']]</td>\n",
       "      <td>92</td>\n",
       "      <td>69</td>\n",
       "      <td>[[1, 6, 'D'], [7, 68, 'S'], [69, 69, 'D']]</td>\n",
       "      <td>BLN1I0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>Dickeya chrysanthemi (Pectobacterium chrysanth...</td>\n",
       "      <td>556</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>327</td>\n",
       "      <td>P06608</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[22, 22...</td>\n",
       "      <td>[[22, 24, 'C'], [25, 38, 'S'], [39, 55, 'C'], ...</td>\n",
       "      <td>348</td>\n",
       "      <td>327</td>\n",
       "      <td>[[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...</td>\n",
       "      <td>II3V73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>Escherichia coli (strain K12)</td>\n",
       "      <td>83333</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>429</td>\n",
       "      <td>P12995</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 429...</td>\n",
       "      <td>[[1, 159, 'S'], [160, 168, 'C'], [169, 182, 'S...</td>\n",
       "      <td>429</td>\n",
       "      <td>429</td>\n",
       "      <td>[[1, 159, 'S'], [160, 168, 'C'], [169, 182, 'S...</td>\n",
       "      <td>JRVQ9Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>Mesobuthus martensii (Manchurian scorpion) (Bu...</td>\n",
       "      <td>34649</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...</td>\n",
       "      <td>64</td>\n",
       "      <td>P45697</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[20, 83...</td>\n",
       "      <td>[[20, 83, 'S']]</td>\n",
       "      <td>84</td>\n",
       "      <td>64</td>\n",
       "      <td>[[1, 64, 'S']]</td>\n",
       "      <td>XUTWNJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>Aquifex aeolicus (strain VF5)</td>\n",
       "      <td>224324</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>216</td>\n",
       "      <td>O67023</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[2, 71,...</td>\n",
       "      <td>[[1, 216, 'S']]</td>\n",
       "      <td>216</td>\n",
       "      <td>216</td>\n",
       "      <td>[[1, 216, 'S']]</td>\n",
       "      <td>3I39SJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>Enterococcus faecalis (strain ATCC 700802 / V583)</td>\n",
       "      <td>226185</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>217</td>\n",
       "      <td>Q82ZI8</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[3, 15,...</td>\n",
       "      <td>[[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "      <td>[[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]</td>\n",
       "      <td>3B4BSR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>Sphingomonas elodea</td>\n",
       "      <td>179878</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>289</td>\n",
       "      <td>Q5FYV5</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 201...</td>\n",
       "      <td>[[1, 1, 'C'], [2, 161, 'S'], [162, 182, 'C'], ...</td>\n",
       "      <td>289</td>\n",
       "      <td>289</td>\n",
       "      <td>[[1, 1, 'C'], [2, 161, 'S'], [162, 182, 'C'], ...</td>\n",
       "      <td>EMRWSZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>9606</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 ...</td>\n",
       "      <td>445</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[263, 2...</td>\n",
       "      <td>[[263, 277, 'C'], [278, 296, 'S'], [297, 297, ...</td>\n",
       "      <td>707</td>\n",
       "      <td>445</td>\n",
       "      <td>[[1, 15, 'C'], [16, 34, 'S'], [35, 35, 'C'], [...</td>\n",
       "      <td>A9X55V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>Salmonella typhimurium (strain LT2 / SGSC1412 ...</td>\n",
       "      <td>99287</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>144</td>\n",
       "      <td>P67091</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[2, 3, ...</td>\n",
       "      <td>[[1, 1, 'D'], [2, 3, 'C'], [4, 13, 'S'], [14, ...</td>\n",
       "      <td>144</td>\n",
       "      <td>144</td>\n",
       "      <td>[[1, 1, 'D'], [2, 3, 'C'], [4, 13, 'S'], [14, ...</td>\n",
       "      <td>YSJNTQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>Arabidopsis thaliana (Mouse-ear cress)</td>\n",
       "      <td>3702</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>207</td>\n",
       "      <td>Q9CA59</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[19, 92...</td>\n",
       "      <td>[[19, 92, 'S'], [93, 99, 'C'], [100, 100, 'S']...</td>\n",
       "      <td>256</td>\n",
       "      <td>207</td>\n",
       "      <td>[[1, 74, 'S'], [75, 81, 'C'], [82, 82, 'S'], [...</td>\n",
       "      <td>1I844M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>9606</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>173</td>\n",
       "      <td>O75628</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[79, 13...</td>\n",
       "      <td>[[79, 102, 'S'], [103, 110, 'C'], [111, 136, '...</td>\n",
       "      <td>298</td>\n",
       "      <td>173</td>\n",
       "      <td>[[1, 24, 'S'], [25, 32, 'C'], [33, 58, 'S'], [...</td>\n",
       "      <td>H7MF8D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>9606</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>350</td>\n",
       "      <td>P19784</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 2, ...</td>\n",
       "      <td>[[1, 9, 'C'], [10, 47, 'S'], [48, 52, 'C'], [5...</td>\n",
       "      <td>350</td>\n",
       "      <td>350</td>\n",
       "      <td>[[1, 9, 'C'], [10, 47, 'S'], [48, 52, 'C'], [5...</td>\n",
       "      <td>7926YW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>Entamoeba histolytica</td>\n",
       "      <td>5759</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>305</td>\n",
       "      <td>Q9U8X2</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 265...</td>\n",
       "      <td>[[1, 211, 'S'], [212, 214, 'C'], [215, 265, 'S...</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "      <td>[[1, 211, 'S'], [212, 214, 'C'], [215, 265, 'S...</td>\n",
       "      <td>7EBEDT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>Mycobacterium paratuberculosis (strain ATCC BA...</td>\n",
       "      <td>262316</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...</td>\n",
       "      <td>136</td>\n",
       "      <td>Q740Y8</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[109, 2...</td>\n",
       "      <td>[[109, 244, 'S']]</td>\n",
       "      <td>244</td>\n",
       "      <td>136</td>\n",
       "      <td>[[1, 136, 'S']]</td>\n",
       "      <td>SUB149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>Salmonella typhimurium (strain LT2 / SGSC1412 ...</td>\n",
       "      <td>99287</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>271</td>\n",
       "      <td>Q8ZK48</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[1, 180...</td>\n",
       "      <td>[[1, 180, 'S'], [181, 181, 'C'], [182, 268, 'S...</td>\n",
       "      <td>271</td>\n",
       "      <td>271</td>\n",
       "      <td>[[1, 180, 'S'], [181, 181, 'C'], [182, 268, 'S...</td>\n",
       "      <td>RYJQNO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>19</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>10090</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>130</td>\n",
       "      <td>P70429</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[2, 13,...</td>\n",
       "      <td>[[1, 27, 'S'], [28, 32, 'D'], [33, 113, 'S'], ...</td>\n",
       "      <td>414</td>\n",
       "      <td>130</td>\n",
       "      <td>[[1, 27, 'S'], [28, 32, 'D'], [33, 113, 'S'], ...</td>\n",
       "      <td>9G0ORR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>Mycobacterium tuberculosis (strain ATCC 25618 ...</td>\n",
       "      <td>83332</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 ...</td>\n",
       "      <td>357</td>\n",
       "      <td>P9WI99</td>\n",
       "      <td>{'disorder': {'derived': [{'regions': [[22, 19...</td>\n",
       "      <td>[[22, 190, 'S'], [191, 195, 'D'], [196, 352, '...</td>\n",
       "      <td>378</td>\n",
       "      <td>357</td>\n",
       "      <td>[[1, 169, 'S'], [170, 174, 'D'], [175, 331, 'S...</td>\n",
       "      <td>AEL3YJ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  Correct Index  \\\n",
       "0            0              0   \n",
       "1            1              1   \n",
       "2            2              2   \n",
       "3            3              4   \n",
       "4            4              5   \n",
       "5            5              6   \n",
       "6            6              7   \n",
       "7            7              8   \n",
       "8            8              9   \n",
       "9            9             10   \n",
       "10          10             11   \n",
       "11          11             12   \n",
       "12          12             13   \n",
       "13          13             14   \n",
       "14          14             15   \n",
       "15          15             16   \n",
       "16          16             17   \n",
       "17          17             18   \n",
       "18          18             19   \n",
       "19          19             20   \n",
       "\n",
       "                                             organism  ncbi_taxon_id  \\\n",
       "0                                Homo sapiens (Human)           9606   \n",
       "1   Kluyveromyces lactis (strain ATCC 8585 / CBS 2...         284590   \n",
       "2    Kluyveromyces lactis (Yeast) (Candida sphaerica)          28985   \n",
       "3                                Mus musculus (Mouse)          10090   \n",
       "4   Dickeya chrysanthemi (Pectobacterium chrysanth...            556   \n",
       "5                       Escherichia coli (strain K12)          83333   \n",
       "6   Mesobuthus martensii (Manchurian scorpion) (Bu...          34649   \n",
       "7                       Aquifex aeolicus (strain VF5)         224324   \n",
       "8   Enterococcus faecalis (strain ATCC 700802 / V583)         226185   \n",
       "9                                 Sphingomonas elodea         179878   \n",
       "10                               Homo sapiens (Human)           9606   \n",
       "11  Salmonella typhimurium (strain LT2 / SGSC1412 ...          99287   \n",
       "12             Arabidopsis thaliana (Mouse-ear cress)           3702   \n",
       "13                               Homo sapiens (Human)           9606   \n",
       "14                               Homo sapiens (Human)           9606   \n",
       "15                              Entamoeba histolytica           5759   \n",
       "16  Mycobacterium paratuberculosis (strain ATCC BA...         262316   \n",
       "17  Salmonella typhimurium (strain LT2 / SGSC1412 ...          99287   \n",
       "18                               Mus musculus (Mouse)          10090   \n",
       "19  Mycobacterium tuberculosis (strain ATCC 25618 ...          83332   \n",
       "\n",
       "                                   shortened_sequence  Sequence_Length  \\\n",
       "0   1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 ...              173   \n",
       "1   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              145   \n",
       "2   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              145   \n",
       "3   1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...               69   \n",
       "4   1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...              327   \n",
       "5   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              429   \n",
       "6   0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...               64   \n",
       "7   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              216   \n",
       "8   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              217   \n",
       "9   0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              289   \n",
       "10  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 ...              445   \n",
       "11  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              144   \n",
       "12  0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 ...              207   \n",
       "13  1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...              173   \n",
       "14  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              350   \n",
       "15  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              305   \n",
       "16  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...              136   \n",
       "17  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              271   \n",
       "18  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...              130   \n",
       "19  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 ...              357   \n",
       "\n",
       "       acc                                   mobidb_consensus  \\\n",
       "0   P31994  {'disorder': {'derived': [{'regions': [[46, 46...   \n",
       "1   F2Z602  {'disorder': {'derived': [{'regions': [[2, 143...   \n",
       "2   Q875M3  {'disorder': {'derived': [{'regions': [[1, 145...   \n",
       "3   P10855  {'disorder': {'derived': [{'regions': [[30, 91...   \n",
       "4   P06608  {'disorder': {'derived': [{'regions': [[22, 22...   \n",
       "5   P12995  {'disorder': {'derived': [{'regions': [[1, 429...   \n",
       "6   P45697  {'disorder': {'derived': [{'regions': [[20, 83...   \n",
       "7   O67023  {'disorder': {'derived': [{'regions': [[2, 71,...   \n",
       "8   Q82ZI8  {'disorder': {'derived': [{'regions': [[3, 15,...   \n",
       "9   Q5FYV5  {'disorder': {'derived': [{'regions': [[1, 201...   \n",
       "10  Q969H0  {'disorder': {'derived': [{'regions': [[263, 2...   \n",
       "11  P67091  {'disorder': {'derived': [{'regions': [[2, 3, ...   \n",
       "12  Q9CA59  {'disorder': {'derived': [{'regions': [[19, 92...   \n",
       "13  O75628  {'disorder': {'derived': [{'regions': [[79, 13...   \n",
       "14  P19784  {'disorder': {'derived': [{'regions': [[1, 2, ...   \n",
       "15  Q9U8X2  {'disorder': {'derived': [{'regions': [[1, 265...   \n",
       "16  Q740Y8  {'disorder': {'derived': [{'regions': [[109, 2...   \n",
       "17  Q8ZK48  {'disorder': {'derived': [{'regions': [[1, 180...   \n",
       "18  P70429  {'disorder': {'derived': [{'regions': [[2, 13,...   \n",
       "19  P9WI99  {'disorder': {'derived': [{'regions': [[22, 19...   \n",
       "\n",
       "                           shortened_mobidb_consensus  Sequence Length  \\\n",
       "0   [[45, 45, 'D'], [46, 49, 'C'], [50, 66, 'S'], ...              310   \n",
       "1       [[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]              145   \n",
       "2                                     [[1, 145, 'S']]              145   \n",
       "3       [[24, 29, 'D'], [30, 91, 'S'], [92, 92, 'D']]               92   \n",
       "4   [[22, 24, 'C'], [25, 38, 'S'], [39, 55, 'C'], ...              348   \n",
       "5   [[1, 159, 'S'], [160, 168, 'C'], [169, 182, 'S...              429   \n",
       "6                                     [[20, 83, 'S']]               84   \n",
       "7                                     [[1, 216, 'S']]              216   \n",
       "8       [[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]              217   \n",
       "9   [[1, 1, 'C'], [2, 161, 'S'], [162, 182, 'C'], ...              289   \n",
       "10  [[263, 277, 'C'], [278, 296, 'S'], [297, 297, ...              707   \n",
       "11  [[1, 1, 'D'], [2, 3, 'C'], [4, 13, 'S'], [14, ...              144   \n",
       "12  [[19, 92, 'S'], [93, 99, 'C'], [100, 100, 'S']...              256   \n",
       "13  [[79, 102, 'S'], [103, 110, 'C'], [111, 136, '...              298   \n",
       "14  [[1, 9, 'C'], [10, 47, 'S'], [48, 52, 'C'], [5...              350   \n",
       "15  [[1, 211, 'S'], [212, 214, 'C'], [215, 265, 'S...              305   \n",
       "16                                  [[109, 244, 'S']]              244   \n",
       "17  [[1, 180, 'S'], [181, 181, 'C'], [182, 268, 'S...              271   \n",
       "18  [[1, 27, 'S'], [28, 32, 'D'], [33, 113, 'S'], ...              414   \n",
       "19  [[22, 190, 'S'], [191, 195, 'D'], [196, 352, '...              378   \n",
       "\n",
       "    #_AA_referenced                                    Updated Indexes  \\\n",
       "0               173  [[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...   \n",
       "1               145      [[1, 1, 'D'], [2, 143, 'S'], [144, 145, 'C']]   \n",
       "2               145                                    [[1, 145, 'S']]   \n",
       "3                69         [[1, 6, 'D'], [7, 68, 'S'], [69, 69, 'D']]   \n",
       "4               327  [[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...   \n",
       "5               429  [[1, 159, 'S'], [160, 168, 'C'], [169, 182, 'S...   \n",
       "6                64                                     [[1, 64, 'S']]   \n",
       "7               216                                    [[1, 216, 'S']]   \n",
       "8               217      [[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]   \n",
       "9               289  [[1, 1, 'C'], [2, 161, 'S'], [162, 182, 'C'], ...   \n",
       "10              445  [[1, 15, 'C'], [16, 34, 'S'], [35, 35, 'C'], [...   \n",
       "11              144  [[1, 1, 'D'], [2, 3, 'C'], [4, 13, 'S'], [14, ...   \n",
       "12              207  [[1, 74, 'S'], [75, 81, 'C'], [82, 82, 'S'], [...   \n",
       "13              173  [[1, 24, 'S'], [25, 32, 'C'], [33, 58, 'S'], [...   \n",
       "14              350  [[1, 9, 'C'], [10, 47, 'S'], [48, 52, 'C'], [5...   \n",
       "15              305  [[1, 211, 'S'], [212, 214, 'C'], [215, 265, 'S...   \n",
       "16              136                                    [[1, 136, 'S']]   \n",
       "17              271  [[1, 180, 'S'], [181, 181, 'C'], [182, 268, 'S...   \n",
       "18              130  [[1, 27, 'S'], [28, 32, 'D'], [33, 113, 'S'], ...   \n",
       "19              357  [[1, 169, 'S'], [170, 174, 'D'], [175, 331, 'S...   \n",
       "\n",
       "   Identifier  \n",
       "0      JV1H0X  \n",
       "1      12KUMW  \n",
       "2      IXJPFB  \n",
       "3      BLN1I0  \n",
       "4      II3V73  \n",
       "5      JRVQ9Y  \n",
       "6      XUTWNJ  \n",
       "7      3I39SJ  \n",
       "8      3B4BSR  \n",
       "9      EMRWSZ  \n",
       "10     A9X55V  \n",
       "11     YSJNTQ  \n",
       "12     1I844M  \n",
       "13     H7MF8D  \n",
       "14     7926YW  \n",
       "15     7EBEDT  \n",
       "16     SUB149  \n",
       "17     RYJQNO  \n",
       "18     9G0ORR  \n",
       "19     AEL3YJ  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating new dataframe consisting of relevant columns for output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "testdf = df[['Identifier', 'Sequence_Length', 'shortened_sequence', 'Updated Indexes']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LA4CYW    1\n",
       "ARS2W4    1\n",
       "U1Q2UK    1\n",
       "7C2S3Y    1\n",
       "23WCE9    1\n",
       "         ..\n",
       "LU0VCA    1\n",
       "0USC82    1\n",
       "74HORQ    1\n",
       "V6S6XI    1\n",
       "3F3SI0    1\n",
       "Name: Identifier, Length: 18091, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testdf['Identifier'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "testdf.rename(columns={'Updated Indexes': 'Updated_Indexes'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating Train and Test Sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add every 4th row to test dataset\n",
    "testSetDf = testdf.iloc[::4,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Sequence_Length</th>\n",
       "      <th>shortened_sequence</th>\n",
       "      <th>Updated_Indexes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>JV1H0X</td>\n",
       "      <td>173</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 ...</td>\n",
       "      <td>[[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>II3V73</td>\n",
       "      <td>327</td>\n",
       "      <td>1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>3B4BSR</td>\n",
       "      <td>217</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>1I844M</td>\n",
       "      <td>207</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 74, 'S'], [75, 81, 'C'], [82, 82, 'S'], [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>SUB149</td>\n",
       "      <td>136</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 136, 'S']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18072</td>\n",
       "      <td>LFETHK</td>\n",
       "      <td>134</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 81, 'S'], [82, 87, 'C'], [88, 118, 'S'], ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18076</td>\n",
       "      <td>86EQSL</td>\n",
       "      <td>21</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 21, 'S']]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18080</td>\n",
       "      <td>W0I9MV</td>\n",
       "      <td>48</td>\n",
       "      <td>0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 3, 'C'], [4, 28, 'S'], [29, 32, 'C'], [33...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18084</td>\n",
       "      <td>0JNLJC</td>\n",
       "      <td>101</td>\n",
       "      <td>0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 9, 'S'], [10, 20, 'C'], [21, 29, 'S'], [3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18088</td>\n",
       "      <td>LF6L1R</td>\n",
       "      <td>169</td>\n",
       "      <td>0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...</td>\n",
       "      <td>[[1, 2, 'C'], [3, 122, 'S'], [123, 129, 'C'], ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4523 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Identifier  Sequence_Length  \\\n",
       "0         JV1H0X              173   \n",
       "4         II3V73              327   \n",
       "8         3B4BSR              217   \n",
       "12        1I844M              207   \n",
       "16        SUB149              136   \n",
       "...          ...              ...   \n",
       "18072     LFETHK              134   \n",
       "18076     86EQSL               21   \n",
       "18080     W0I9MV               48   \n",
       "18084     0JNLJC              101   \n",
       "18088     LF6L1R              169   \n",
       "\n",
       "                                      shortened_sequence  \\\n",
       "0      1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 ...   \n",
       "4      1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "8      0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "12     0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "16     0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 ...   \n",
       "...                                                  ...   \n",
       "18072  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 ...   \n",
       "18076  0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 ...   \n",
       "18080  0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "18084  0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "18088  0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ...   \n",
       "\n",
       "                                         Updated_Indexes  \n",
       "0      [[1, 1, 'D'], [2, 5, 'C'], [6, 22, 'S'], [23, ...  \n",
       "4      [[1, 3, 'C'], [4, 17, 'S'], [18, 34, 'C'], [35...  \n",
       "8          [[1, 2, 'D'], [3, 215, 'S'], [216, 217, 'C']]  \n",
       "12     [[1, 74, 'S'], [75, 81, 'C'], [82, 82, 'S'], [...  \n",
       "16                                       [[1, 136, 'S']]  \n",
       "...                                                  ...  \n",
       "18072  [[1, 81, 'S'], [82, 87, 'C'], [88, 118, 'S'], ...  \n",
       "18076                                     [[1, 21, 'S']]  \n",
       "18080  [[1, 3, 'C'], [4, 28, 'S'], [29, 32, 'C'], [33...  \n",
       "18084  [[1, 9, 'S'], [10, 20, 'C'], [21, 29, 'S'], [3...  \n",
       "18088  [[1, 2, 'C'], [3, 122, 'S'], [123, 129, 'C'], ...  \n",
       "\n",
       "[4523 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testSetDf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add remaining rows to train dataset\n",
    "trainSetDf = pd.concat([testdf, testSetDf, testSetDf]).drop_duplicates(keep=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Formatting test set for output file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restructuring data in order to extract relevant information for text file input for Neural Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_list = []\n",
    "\n",
    "for rows in testSetDf.itertuples(): # iterating through rows in dataframe\n",
    "    testSetIndexes = [rows.Updated_Indexes] # Adding indexes with class to a list\n",
    "    test_list.append(testSetIndexes)\n",
    "    \n",
    "flat_test_list = [] \n",
    "\n",
    "for sublist in test_list:\n",
    "    for item in sublist:\n",
    "        flat_test_list.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting elements in list from string to list\n",
    "testdf_x = [ast.literal_eval(s) for s in flat_test_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "classTypeList = []\n",
    "\n",
    "for value in testdf_x:\n",
    "    classType = ''\n",
    "    for x in value: # Printing corresponding class for each amino acid in the sequence\n",
    "        classType += ((x[1] - x[0]+1) * ((x[2])+' '))\n",
    "    classTypeList.append(classType)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rebeccalenihan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "testSetDf['correspondingClassToAA'] = classTypeList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rebeccalenihan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "/Users/rebeccalenihan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/rebeccalenihan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "testSetDf['correspondingClassToAA'] = testSetDf['correspondingClassToAA'].str.replace('D','1')\n",
    "testSetDf['correspondingClassToAA'] = testSetDf['correspondingClassToAA'].str.replace('S','0')\n",
    "testSetDf['correspondingClassToAA'] = testSetDf['correspondingClassToAA'].str.replace('C','2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing the relevant data to output file in correct format for Neural Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "file=open('test.dataset','a')\n",
    "\n",
    "for idx,row in testSetDf.iterrows(): # iterate through dataset \n",
    "    nnInputTest = str(row['Identifier'])+ '\\n' # first line consists of protein ID\n",
    "    nnInputTest += str(row['Sequence_Length'])+ '\\n' # Second line consists of protein sequence length\n",
    "    for sequence in row['shortened_sequence']: # third line consists of protein sequence\n",
    "        nnInputTest += str(sequence)\n",
    "    nnInputTest += '\\n'\n",
    "    nnInputTest += str(row['correspondingClassToAA'])+ '\\n' # final line consists of class corresponding to each amino acid in sequence    \n",
    "    nnInputTest += '\\n'\n",
    "    file.write(nnInputTest) # write to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Formatting train set for output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_list = []\n",
    "\n",
    "for rows in trainSetDf.itertuples():\n",
    "    trainSetIndexes = [rows.Updated_Indexes]\n",
    "    train_list.append(trainSetIndexes)\n",
    "    \n",
    "flat_train_list = []\n",
    "for sublist in train_list:\n",
    "    for item in sublist:\n",
    "        flat_train_list.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting elements in list from string to list\n",
    "traindf_x = [ast.literal_eval(s) for s in flat_train_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "classTypeList = []\n",
    "\n",
    "for value in traindf_x:\n",
    "    classType = ''\n",
    "    for x in value: # Printing corresponding class for each amino acid in the sequence\n",
    "        classType += ((x[1] - x[0]+1) * ((x[2])+' '))\n",
    "    classTypeList.append(classType)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "trainSetDf['correspondingClassToAA'] = classTypeList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "trainSetDf['correspondingClassToAA'] = trainSetDf['correspondingClassToAA'].str.replace('D','1')\n",
    "trainSetDf['correspondingClassToAA'] = trainSetDf['correspondingClassToAA'].str.replace('S','0')\n",
    "trainSetDf['correspondingClassToAA'] = trainSetDf['correspondingClassToAA'].str.replace('C','2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file=open('train.dataset','a')\n",
    "\n",
    "for idx,row in trainSetDf.iterrows(): # iterate through dataset\n",
    "    nnInputTrain  = str(row['Identifier'])+ '\\n' # first line consists of protein ID\n",
    "    nnInputTrain += str(row['Sequence_Length'])+ '\\n'  # Second line consists of protein sequence length\n",
    "    for sequence in row['shortened_sequence']: # third line consists of protein sequence\n",
    "        nnInputTrain += str(sequence)\n",
    "    nnInputTrain += '\\n'\n",
    "    nnInputTrain += str(row['correspondingClassToAA'])+ '\\n' # final line consists of class corresponding to each amino acid in sequence    \n",
    "    nnInputTrain += '\\n'\n",
    "    file.write(nnInputTrain) # write to file"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
